pkl-core/src/main/java/org/pkl/core/ast/type/VmTypeMismatchException.java

```diff
@@ -35,18 +38,35 @@
 public abstract class VmTypeMismatchException extends ControlFlowException {
   protected final SourceSection sourceSection;
   protected final Object actualValue;
+  protected @Nullable Map<CallTarget, StackFrame> insertedStackFrames = new HashMap<>();
 
   protected VmTypeMismatchException(SourceSection sourceSection, Object actualValue) {
     this.sourceSection = sourceSection;
     this.actualValue = actualValue;
   }
 
+  @TruffleBoundary
+  public void putInsertedStackFrame(CallTarget callTarget, StackFrame stackFrame) {
+    if (this.insertedStackFrames == null) {
+      this.insertedStackFrames = new HashMap<>();
+    }
+    this.insertedStackFrames.put(callTarget, stackFrame);
+  }
```

Comment:
Added this so we can allow a `VmTypeMismatchException` to blame a source section when a typecheck fails (see `ListingOrMappingTypeNode#doEagerCheck`)